File: src/main/java/io/vproxy/vfx/ui/table/VTableColumn.java
package io.vproxy.vfx.ui.table;

import io.vproxy.vfx.manager.font.FontManager;
import io.vproxy.vfx.manager.font.FontUsages;
import io.vproxy.vfx.theme.Theme;
import io.vproxy.vfx.ui.layout.HPadding;
import io.vproxy.vfx.util.FXUtils;
import javafx.geometry.Insets;
import javafx.geometry.Pos;
import javafx.scene.Node;
import javafx.scene.control.Label;
import javafx.scene.layout.*;
import javafx.scene.paint.Color;
import javafx.scene.paint.CycleMethod;
import javafx.scene.paint.LinearGradient;
import javafx.scene.paint.Stop;

import java.util.Comparator;
import java.util.function.Function;

import static io.vproxy.vfx.ui.table.VTableSortOrder.ASC;
import static io.vproxy.vfx.ui.table.VTableSortOrder.DESC;

public class VTableColumn<S, T> {
    private static final Color COLOR_TOP = Theme.current().tableHeaderTopBackgroundColor();// = new Color(0xef / 255d, 0xef / 255d, 0xef / 255d, 1);
    private static final Color COLOR_BOT = Theme.current().tableHeaderBottomBackgroundColor();// = new Color(0xe1 / 255d, 0xe1 / 255d, 0xe1 / 255d, 1);
    static final Background BG = new Background(new BackgroundFill(
        new LinearGradient(
            0, 0, 0, 1, true, CycleMethod.NO_CYCLE,
            new Stop(0, COLOR_TOP),
            new Stop(1, COLOR_BOT)
        ),
        CornerRadii.EMPTY,
        Insets.EMPTY
    ));
    private static final int sortWidth = 15;

    public final String name;
    final Function<S, T> valueRetriever;
    Comparator<T> comparator;
    final HBox columnNode = new HBox();
    private final Label sortLabel;
    Function<T, ? extends Node> nodeBuilder = null;

    VTableSharedData<S> shared;

    final VBox vbox = new VBox();
    double minWidth = 0;
    double maxWidth = 0;
    double prefWidth = 0;
    Pos alignment;

    private int sortPriority = 0; // 0 means no sort
    private VTableSortOrder sortOrder = DESC; // the first time it will be set to ASC

    public VTableColumn(String name, Function<S, T> valueRetriever) {
        this(name, new Label(name) {{
            setAlignment(Pos.CENTER);
            setTextFill(Theme.current().tableHeaderTextColor());
            setPrefHeight(25);
        }}, valueRetriever);
    }

    public VTableColumn(String name, Region columnContentNode, Function<S, T> valueRetriever) {
        this.name = name;
        this.valueRetriever = valueRetriever;
        sortLabel = new Label() {{
            setPadding(new Insets(0, 4, 0, 4));
            setPrefWidth(sortWidth);
            setTextFill(Theme.current().tableSortLabelColor());
        }};
        columnNode.getChildren().addAll(
            new HPadding(sortWidth),
            columnContentNode,
            sortLabel
        );
        columnNode.setBackground(BG);
        FXUtils.observeHeight(columnContentNode, columnNode);
        columnNode.setAlignment(Pos.CENTER);

        columnNode.setOnMouseClicked(e -> {
            if (comparator == null) {
                if (sortPriority != 0) {
                    shared.tableView.clearSort(this);
                }
                return;
            }
            if (shared != null) {
                if (sortPriority != 0 && sortOrder == DESC) {
                    shared.tableView.clearSort(this);
                } else {
                    shared.tableView.sortBy(this, sortOrder == ASC ? DESC : ASC);
                }
            }
        });

        columnNode.widthProperty().addListener((ob, old, now) -> {
            if (now == null) return;
            columnContentNode.setPrefWidth(now.doubleValue() - 2 * sortWidth);
        });
        vbox.widthProperty().addListener((ob, old, now) -> {
            if (now == null) return;
            columnNode.setPrefWidth(now.doubleValue());
            columnNode.setMinWidth(now.doubleValue());
            columnNode.setMaxWidth(now.doubleValue());
        });
    }

    public void setComparator(Comparator<T> comparator) {
        this.comparator = comparator;
        if (sortPriority > 0 && shared != null) {
            if (comparator == null) {
                shared.tableView.clearSort(this);
            } else {
                shared.tableView.sort();
            }
        }
    }

    public void setMinWidth(double minWidth) {
        this.minWidth = minWidth;
        if (shared != null) {
            shared.tableView.updateWidth();
        }
    }

    public void setMaxWidth(double maxWidth) {
        this.maxWidth = maxWidth;
        if (shared != null) {
            shared.tableView.updateWidth();
        }
    }

    public void setPrefWidth(double prefWidth) {
        this.prefWidth = prefWidth;
        if (shared != null) {
            shared.tableView.updateWidth();
        }
    }

    public void setAlignment(Pos alignment) {
        this.alignment = alignment;
        if (shared != null) {
            shared.tableView.updateRowNodeForColumn(this);
        }
    }

    public void setTextBuilder(Function<T, String> textBuilder) {
        setNodeBuilder(t -> {
            var str = textBuilder.apply(t);
            return new Label(str) {{
                setTextFill(Theme.current().tableTextColor());
                FontManager.get().setFont(FontUsages.tableCellText, this);
            }};
        });
    }

    public void setNodeBuilder(Function<T, ? extends Node> nodeBuilder) {
        this.nodeBuilder = nodeBuilder;
        if (shared != null) {
            shared.tableView.refresh();
        }
    }

    void initCell(VTableCellPane<S> cell) {
        if (alignment != null) {
            cell.setAlignment(alignment);
        }
    }

    public int getSortPriority() {
        return sortPriority;
    }

    public VTableSortOrder getSortOrder() {
        return sortOrder;
    }

    void setSort(int priority, VTableSortOrder order) {
        this.sortPriority = priority;
        this.sortOrder = order;
        updateSortLabel();
    }

    private void updateSortLabel() {
        if (sortPriority == 0) {
            sortLabel.setText("");
            return;
        }
        if (sortOrder == ASC) {
            sortLabel.setRotate(0);
            sortLabel.setAlignment(Pos.CENTER_RIGHT);
        } else {
            sortLabel.setRotate(180);
            sortLabel.setAlignment(Pos.CENTER_LEFT);
        }
        sortLabel.setText("" + sortPriority);
    }

    void resetSortOrder() {
        this.sortOrder = DESC;
    }

    void resetSortPriority() {
        sortPriority = 0;
        sortLabel.setText("");
    }

    void decSortPriority() {
        --this.sortPriority;
        updateSortLabel();
    }
}

File: src-ts/src/ui/table/VTableColumn.ts
import { Node } from '../../javafx/Node.js';
import { Label } from '../../javafx/Label.js';
import { HBox } from '../../javafx/HBox.js';
import { VBox } from '../../javafx/VBox.js';
import {
  Background,
  BackgroundFill,
  CornerRadii,
  Insets,
  Pos,
} from '../../javafx/layout.js';
import { LinearGradient } from '../../javafx/color.js';
import { FontManager } from '../../manager/font/FontManager.js';
import { FontUsages } from '../../manager/font/FontUsages.js';
import { Theme } from '../../theme/Theme.js';
import { FXUtils } from '../../util/FXUtils.js';
import { HPadding } from '../layout/HPadding.js';
import { VTableSortOrder } from './VTableSortOrder.js';
import type { VTableCellPane } from './VTableCellPane.js';
import type { VTableSharedData } from './VTableSharedData.js';

type Region = Node;

export class VTableColumn<S, T> {
  private static _BG: Background | null = null;
  static get BG(): Background {
    if (VTableColumn._BG === null) {
      VTableColumn._BG = new Background(
        new BackgroundFill(
          new LinearGradient(0, 0, 0, 1, true, 'NO_CYCLE', [
            { offset: 0, color: Theme.current().tableHeaderTopBackgroundColor() },
            { offset: 1, color: Theme.current().tableHeaderBottomBackgroundColor() },
          ]),
          CornerRadii.EMPTY,
          Insets.EMPTY,
        ),
      );
    }
    return VTableColumn._BG;
  }
  private static readonly sortWidth = 15;

  public readonly name: string;
  readonly valueRetriever: (s: S) => T;
  comparator: ((a: T, b: T) => number) | null = null;
  readonly columnNode: HBox = new HBox();
  private sortLabel!: Label;
  nodeBuilder: ((t: T) => Node) | null = null;

  shared: VTableSharedData<S> | null = null;

  readonly vbox: VBox = new VBox();
  minWidth: number = 0;
  maxWidth: number = 0;
  prefWidth: number = 0;
  alignment: Pos | null = null;

  private sortPriority: number = 0;
  private sortOrder: VTableSortOrder = VTableSortOrder.DESC;

  constructor(name: string, valueRetriever: (s: S) => T);
  constructor(name: string, columnContentNode: Region, valueRetriever: (s: S) => T);
  constructor(name: string, columnContentNodeOrRetriever: Region | ((s: S) => T), valueRetrieverMaybe?: (s: S) => T) {
    this.name = name;
    if (typeof columnContentNodeOrRetriever === 'function') {
      const valueRetriever = columnContentNodeOrRetriever as (s: S) => T;
      this.valueRetriever = valueRetriever;
      const headerLabel = new Label(name);
      headerLabel.setAlignment(Pos.CENTER);
      headerLabel.setTextFill(Theme.current().tableHeaderTextColor());
      headerLabel.setPrefHeight(25);
      this.finishCtor(headerLabel);
    } else {
      const columnContentNode = columnContentNodeOrRetriever as Region;
      this.valueRetriever = valueRetrieverMaybe!;
      this.finishCtor(columnContentNode);
    }
  }

  private finishCtor(columnContentNode: Region): void {
    this.sortLabel = new Label();
    this.sortLabel.setPadding(new Insets(0, 4, 0, 4));
    this.sortLabel.setPrefWidth(VTableColumn.sortWidth);
    this.sortLabel.setTextFill(Theme.current().tableSortLabelColor());

    this.columnNode.getChildren().addAll(
      new HPadding(VTableColumn.sortWidth),
      columnContentNode,
      this.sortLabel,
    );
    this.columnNode.setBackground(VTableColumn.BG);
    FXUtils.observeHeight(columnContentNode as unknown as import('../../javafx/Parent.js').Parent, this.columnNode);
    this.columnNode.setAlignment(Pos.CENTER);

    this.columnNode.setOnMouseClicked(() => {
      if (this.comparator === null) {
        if (this.sortPriority !== 0) {
          this.shared!.tableView.clearSort(this as VTableColumn<S, unknown>);
        }
        return;
      }
      if (this.shared !== null) {
        if (this.sortPriority !== 0 && this.sortOrder === VTableSortOrder.DESC) {
          this.shared.tableView.clearSort(this as VTableColumn<S, unknown>);
        } else {
          this.shared.tableView.sortBy(this as VTableColumn<S, unknown>, this.sortOrder === VTableSortOrder.ASC ? VTableSortOrder.DESC : VTableSortOrder.ASC);
        }
      }
    });

    this.columnNode.widthProperty.addListener((_old, now) => {
      if (now === null) return;
      columnContentNode.setPrefWidth(now - 2 * VTableColumn.sortWidth);
    });
    this.vbox.widthProperty.addListener((_old, now) => {
      if (now === null) return;
      this.columnNode.setPrefWidth(now);
      this.columnNode.setMinWidth(now);
      this.columnNode.setMaxWidth(now);
    });
  }

  setComparator(comparator: ((a: T, b: T) => number) | null): void {
    this.comparator = comparator;
    if (this.sortPriority > 0 && this.shared !== null) {
      if (comparator === null) {
        this.shared.tableView.clearSort(this as VTableColumn<S, unknown>);
      } else {
        this.shared.tableView.sort();
      }
    }
  }

  setMinWidth(minWidth: number): void {
    this.minWidth = minWidth;
    if (this.shared !== null) {
      this.shared.tableView.updateWidth();
    }
  }

  setMaxWidth(maxWidth: number): void {
    this.maxWidth = maxWidth;
    if (this.shared !== null) {
      this.shared.tableView.updateWidth();
    }
  }

  setPrefWidth(prefWidth: number): void {
    this.prefWidth = prefWidth;
    if (this.shared !== null) {
      this.shared.tableView.updateWidth();
    }
  }

  setAlignment(alignment: Pos): void {
    this.alignment = alignment;
    if (this.shared !== null) {
      this.shared.tableView.updateRowNodeForColumn(this as VTableColumn<S, unknown>);
    }
  }

  setTextBuilder(textBuilder: (t: T) => string): void {
    this.setNodeBuilder((t) => {
      const str = textBuilder(t);
      const label = new Label(str);
      label.setTextFill(Theme.current().tableTextColor());
      FontManager.get().setFont(FontUsages.tableCellText, label);
      return label;
    });
  }

  setNodeBuilder(nodeBuilder: ((t: T) => Node) | null): void {
    this.nodeBuilder = nodeBuilder;
    if (this.shared !== null) {
      this.shared.tableView.refresh();
    }
  }

  initCell(cell: VTableCellPane<S>): void {
    if (this.alignment !== null) {
      cell.setAlignment(this.alignment);
    }
  }

  getSortPriority(): number {
    return this.sortPriority;
  }

  getSortOrder(): VTableSortOrder {
    return this.sortOrder;
  }

  setSort(priority: number, order: VTableSortOrder): void {
    this.sortPriority = priority;
    this.sortOrder = order;
    this.updateSortLabel();
  }

  private updateSortLabel(): void {
    if (this.sortPriority === 0) {
      this.sortLabel.setText('');
      this.sortLabel.el.style.transform = '';
      return;
    }
    if (this.sortOrder === VTableSortOrder.ASC) {
      this.sortLabel.el.style.transform = 'rotate(0deg)';
      this.sortLabel.setAlignment(Pos.CENTER_RIGHT);
    } else {
      this.sortLabel.el.style.transform = 'rotate(180deg)';
      this.sortLabel.setAlignment(Pos.CENTER_LEFT);
    }
    this.sortLabel.setText('' + this.sortPriority);
  }

  resetSortOrder(): void {
    this.sortOrder = VTableSortOrder.DESC;
  }

  resetSortPriority(): void {
    this.sortPriority = 0;
    this.sortLabel.setText('');
    this.sortLabel.el.style.transform = '';
  }

  decSortPriority(): void {
    --this.sortPriority;
    this.updateSortLabel();
  }
}
